webplugin/js/app/kommunicate-ui.js

```diff
@@ -1803,4 +1803,220 @@ KommunicateUI = {
+                var timeRemaining = endTime - now;
+                var percent = timeElapsed / MAX_RECORD_TIME * 100;
+                var secondsElapsed = Math.floor((timeElapsed / 1000) % 60);
+                var secondsRemaining = Math.floor((timeRemaining / 1000) % 60);
+                var minutesElapsed = Math.floor((timeElapsed / (1000 * 60)) % 60);
+                var minutesRemaining = Math.floor((timeRemaining / (1000 * 60)) % 60);
+                timeElapsedTimer.text(("0" + minutesElapsed).slice(-2) + ":" + ("0" + secondsElapsed).slice(-2));
+                timeRemainingTimer.text(("0" + minutesRemaining).slice(-2) + ":" + ("0" + secondsRemaining).slice(-2));
+                $("#wave-front-progressBar").width(percent+'%');
+                if (timeRemaining <= 0) {
+                    stopRecording();
+                }
+            }, 1000);
+        };
+        function stopRecording(e) {
+            // show
+            kommunicateCommons.modifyClassList(
+                {
+                    id: [
+                        'delete-recording',
+                        'play-btn',
+                        'send-btn',
+                        'time-remaining'
+                    ],
+                },
+                '',
+                'n-vis'
+            );
+            
+            // hide
+            kommunicateCommons.modifyClassList(
+                {
+                    id: [
+                        'pause-btn',
+                        'mck-stop-recording',
+                        'time-elapsed'
+                    ],
+                },
+                'n-vis',
+                ''
+            );
+            clearInterval(recorderInterval);
+            $("#wave-front-progressBar").width('0%');
+            timeRemainingTimer.text("00:00");
+            Fr.voice.pause();
+            Fr.voice.export(function (blob) {
+                blob.name = 'voiceRecord-' + new Date().getTime() + '.wav';
+                params.file = blob;
+                params.callback = function(){
+                    document.querySelector("#send-btn").classList.remove('disabled');
+                }
+                audioBlob = blob;
+                $applozic.fn.applozic('audioAttach', params);
+                var bloburl = URL.createObjectURL(audioBlob);
+                $("#recorder-audio").attr("src", bloburl);
+            }, 'blob');
+            Fr.voice.stop();
+            
+        };
+        function resetRecorder(e, permissionDenied) {
+            Kommunicate.typingAreaService.hideRecorder();
+            !permissionDenied && Fr.voice.stop();
+            clearInterval(recorderInterval);
+            clearInterval(playPauseInterval);
+            recorderInterval = null;
+            audioBlob = null;
+            wavAudioDuration = null;
+            recorderAudio.setAttribute('src', '');
+            isTimerPaused = false;
+            hasTimerExpired = true;
```

Comment:
are these related? is it possible isTimerPaused and hasTimerExpired are both true or both false at the same time? or they both are always going to be opposite of each other?

Reply:
if the timer is running then `isTimerPaused = false` and also `hasTimerExpired = false`.
if the timer is paused then `isTimerPaused = true` and also `hasTimerExpired = false`.
if the timer is cleared then `isTimerPaused = false` and also `hasTimerExpired = true`.